### DISCUSSION TITLE
How exactly are conflicting changes resolved between clones?

### DISCUSSION THREAD

**Discussion by mcalligator** [m-ld/m-ld-spec#99] [Discussion]
A collaborator on a project asked (in another channel):

'how do you achieve convergence when multiple (possibly conflicting) edits of the structured data come in different order in remote nodes?

'I am mostly interested in this statement from [Concurrency topic in the docs](https://www.m-ld.org/doc/#concurrency) "Transaction operations combine to realise the final convergent state of every clone" - how do these transactions "combine"?'

**Comment by gsvarovsky** [comment]
This question gets asked a lot! We need to get it into the FAQs or the docs on the website.

When a user is editing information in an app, it's immediately committed to the local clone and published to other clones. So, if more than one user edits information in the same domain, they could edit the same data at the same time. **m-ld** [guarantees](https://m-ld.org/doc/#concurrency) that both users will eventually see the same data. But how does it merge the edits, without conflicts?

Because the information in **m-ld** is actually a [graph](https://m-ld.org/doc/#structured-data), the most basic behaviour is a 'graph merge'. Imagine that every JSON property entered by a user is an "edge", going from the JSON object itself, to the JSON value. _All_ of the information in the domain is represented like this, so, the **m-ld** domain is actually a _Set_ of graph edges.

In an update, the app might _delete_ some edges and _insert_ some others (this is why [Updates](https://spec.m-ld.org/#transactions) have the structure they do). The algorithm for the 'graph merge' is published computer science – an [SU-Set CRDT](https://hal.inria.fr/hal-00686484/document) – but its effects are simple: inserting an edge wins over deleting the same edge; and otherwise the order doesn't matter and all changes are applied.

This strategy is simple, but it can give rise to some effects which an app has to deal with. For example, if one user changes a JSON property to `"name": "Fred"`, and at the same time another changes it to `"name": "Freddy"`, the final state in the domain is `"name": ["Fred", "Freddy"]` – two graph edges for the same JSON property.

At this point, it's natural to think this should be magically resolved. However, first of all, in many apps it would be good to show _both_ values to the user; after all, this situation was caused by a human interaction, it makes sense to have another human interaction to resolve it.

Another option is for the app to have a sensible default in its _presentation_, like picking the highest value, without actually resolving the conflict in the domain. In that case one user might see that they have 'lost', and re-do their change, collapsing the two values back to one; but if not, it might be good to know later that a conflict happened and no-one dealt with it. In highly interactive apps these options give the users autonomy instead of letting the machine make all the decisions; that's why the default is _not_ to resolve every "conflict".

Of course there are situations in which the machine can make a good, obvious decision, and it would simplify app development if the data were guaranteed to always conform to some data _schema_ – like a certain field must only ever have one value. For this, **m-ld** has a way to extend the domain with [constraints](https://spec.m-ld.org/#constraints). We've shown that constraints can be used to implement quite sophisticated rules (including [embedding other CRDTs](https://m-ld.org/news/#truth-and-just-lists)). We're currently experimenting with improved ways to declare them using a schema language, and we'd [welcome your input](https://github.com/m-ld/m-ld-spec/issues/73).

Finally, there are situations in which conflict resolution is not possible without some kind of coordination between clones. This is necessary for security, and for changes to the schema itself. Our [active research](https://github.com/m-ld/m-ld-security-spec) into this will lead to the ability to integrate with – or supplant – other highly-coordinated information systems, like blockchains.

There's lots more we could talk about like text support (#35) and git-style branch & merge (#68). In the meantime what do you think?

**Reply by kroky** [reply]
This makes a lot of sense - leaving the resolution to the app when conflicts cannot be automatically resolved. When the app (user) decides which value is the correct one to stay (like `Freddy` for example), does it send another update to the clone which effectively changes the `name` edge from `["Fred", "Freddy"]` back to `Freddy`?

Having text merge support out of the box in m-ld might mean even tighter integration with Tiki as we have been considering different approaches for years here but never really shipped with something stable.

**Reply by gsvarovsky** [reply]
> does it send another update to the clone

Yes, exactly. In this case the app could explicitly delete the wrong prior value:

```json
{
  "@delete": { "@id": "fred", "name": "Fred" }
}
```

or just remove any previous values with a wildcard and insert the right one:

```json
{
  "@delete": { "@id": "fred", "name": "?" },
  "@insert": { "@id": "fred", "name": "Freddy" }
}
```

**Reply by gsvarovsky** [reply]
- #100 

😇
